12-offer-encoding.md

```diff
@@ -0,0 +1,769 @@
+it.  The merchant can claim an invoice was paid, and once revealed, anyone can
+claim they paid the invoice, too.[1]
+
+Providing a key in *invoice_request* allows a user to prove that they were the one
+to request the invoice.  In addition, the Merkle construction of the BOLT 12
+invoice signature allows the user to selectively reveal fields of the invoice
+in case of dispute.
+
+# Encoding
+
+Each of the forms documented here are in
+[TLV](01-messaging.md#type-length-value-format) format.
+
+The supported ASCII encoding is the human-readable prefix, followed by a
+`1`, followed by a bech32-style data string of the TLVs in order,
+optionally interspersed with `+` (for indicating additional data is to
+come).
+
+## Requirements
+
+Readers of a bolt12 string:
+- if it encounters a `+` followed zero or more whitespace characters between 
+  two bech32 characters:
+  - MUST remove the `+` and whitespace.
+
+## Rationale
+
+The use of bech32 is arbitrary, but already exists in the bitcoin
+world.  We currently omit the six-character trailing checksum: QR
+codes have their own checksums, bech32 doesn't protect against many
+length differences, and bech32m is not yet widely supported.
+
+The use of `+` (which is ignored) allows use over limited
+text fields like Twitter:
+
+```
+lno1xxxxxxxx+
+
+yyyyyyyyyyyy+
+
+zzzzz
+```
+
+See [format-string-test.json](bolt12/format-string-test.json).
+
+## Signature Calculation
+
+All signatures are created as per
+[BIP-340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki),
+and tagged as recommended there.  Thus we define H(`tag`,`msg`) as
+SHA256(SHA256(`tag`) || SHA256(`tag`) || `msg`), and SIG(`tag`,`msg`,`key`)
+as the signature of H(`tag`,`msg`) using `key`.
+
+Each form is signed using one or more TLV signature elements; TLV
+types 240 through 1000 are considered signature elements.  For these
+the tag is "lightning" || `messagename` || `fieldname`, and `msg` is the
+Merkle-root; "lightning" is the literal 9-byte ASCII string,
+`messagename` is the name of the TLV stream being signed (i.e. "offer", "invoice_request" or "invoice") and the `fieldname` is the TLV field containing the
+signature (e.g. "signature" or "payer_signature").
+
+The formulation of the Merkle tree is similar to that proposed in
+[BIP-taproot], with each TLV leaf paired with a nonce leaf to avoid
+revealing adjacent nodes in proofs (assuming there is a non-revealed TLV
+which has enough entropy).
+
+The Merkle tree's leaves are, in TLV-ascending order for each tlv:
+1. The H(`LnLeaf`,tlv).
+2. The H(`LnAll`||all-tlvs,tlv) where "all-tlvs" consists of all non-signature TLV entries appended in ascending order.
+
+The Merkle tree inner nodes are H(`LnBranch`, lesser-SHA256||greater-SHA256);
+this ordering means that proofs are more compact since left/right is
+inherently determined.
+
+If there are not exactly a power of 2 leaves, then the tree depth will
+be uneven, with the deepest tree on the lowest-order leaves.
```

Comment:
No;  if there are 8 leaves they'll all have depth 3, if there are 5 then 4 have depth 3, one has depth 1.

(Note the above statement is true, but of course we can't have an odd number of leaves in this construction: if there are 8 TLVs, there are 16 leaves, since each gets a nonce-leaf next to it).

Reply:
Ah, I didn't make it clear that I meant only in the case that there is an odd number of TLV fields. Also I think you are _not_ counting the hashing of a leave with its nonce as the first depth? 

Isn't it then the case that if there is an odd number of TLV fields (L1, L2, ..., LN with N being odd), then LNA (the highest-order leave) is simply "carried" up until it hashes with the concatenation of the rest of hashes to form the root? For for the odd case the highest-order leave always has a depth of 1 (or 2 if hashing with its nonce counts as a depth)? This is different from the merkle tree of the transactions in a Bitcoin block, where the last transaction is duplicated and paired with itself as far as I know.

Reply:
Yes.  You should never implement merkle trees like Bitcoin!